src/main/java/org/apache/mesos/scheduler/recovery/DefaultRecoveryPlanManager.java

```diff
@@ -25,93 +23,68 @@
  * {@code Plan}. {@link DefaultRecoveryPlanManager} tracks currently failed (permanent) and stopped (transient) tasks,
  * generates a new {@link DefaultRecoveryBlock} for them and adds them to the recovery Plan, if not already added.
  */
-public class DefaultRecoveryPlanManager implements PlanManager {
+public class DefaultRecoveryPlanManager extends ChainedObserver implements PlanManager {
     private static final Logger LOGGER = LoggerFactory.getLogger(DefaultRecoveryPlanManager.class);
 
     protected volatile Plan plan;
     protected volatile PhaseStrategy phaseStrategy;
     protected volatile Map<UUID, PhaseStrategy> phaseStrategies;
 
     private final StateStore stateStore;
-    private final TaskFailureListener failureListener;
     private final RecoveryRequirementProvider offerReqProvider;
     private final FailureMonitor failureMonitor;
     private final LaunchConstrainer launchConstrainer;
     private final UUID phaseId = UUID.randomUUID();
 
     public DefaultRecoveryPlanManager(
             StateStore stateStore,
-            TaskFailureListener failureListener,
             RecoveryRequirementProvider offerReqProvider,
             LaunchConstrainer launchConstrainer,
             FailureMonitor failureMonitor) {
         this.stateStore = stateStore;
         this.offerReqProvider = offerReqProvider;
         this.failureMonitor = failureMonitor;
         this.launchConstrainer = launchConstrainer;
-        this.failureListener = failureListener;
+        updatePlan(new ArrayList<>());
```

Comment:
Nit: s/`new ArrayList<>()`/`Collections.EMPTY_LIST`/g (feel free to ignore this one)